HPCC-25498 Check and send error messages in DFUFileAccessV2

wangkx opened the pull request:
1. Check the keyPairName setting in /EnvSettings/Keys/ClusterGroup.
2. Check the privateKey setting in /EnvSettings/Keys/KeyPair.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-14890, label: h25498 starts now.
The reason of this test is: Forced to re-test.
Commit ID: 0f3413e21b83b6c769e67bdbec9e6e08ee49722c
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-52.ca-central-1.compute.internal

jakesmith commented on esp/clients/wsdfuaccess/wsdfuaccess.cpp:
```diff
@@ -525,6 +525,8 @@ StringBuffer &encodeDFUFileMeta(StringBuffer &metaInfoBlob, IPropertyTree *metaI
         const char *keyPairName = metaInfo->queryProp("keyPairName");
         const char *privateKeyFName = environment->getPrivateKeyPath(keyPairName);
+        if (isEmptyString(privateKeyFName))
+            throw makeStringExceptionV(-1, "Key name %s is not found in environment settings: /EnvSettings/Keys/KeyPair.", keyPairName);
```
trivial: perhaps change to '%s' - it can sometimes be confusing depending on naming if unquoted.

wangkx replied:
@jake I just made the change.

jakesmith reviewed (commented):
@wangkx - looks fine, 1 trivial issue, please address and squash.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-52.ca-central-1.compute.internal
Sha: 0f3413e21b83b6c769e67bdbec9e6e08ee49722c
Build: success
Build: failed!
Number of Error(s): 12
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:49 (file):
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:56 (file):
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:63 (file):
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:70 (file):
CPack Error: Error when generating package: hpccsystems-platform
make: *** [Makefile:130: package] Error 1
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:49 (file):
  file RPATH_CHANGE could not write new RPATH:
    /opt/HPCCSystems/lib
  to the file:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-core/libaws-cpp-sdk-core.so
  No valid ELF RPATH or RUNPATH entry exists in the file;
Call Stack (most recent call first):
  /home/centos/smoketest/PR-14890/build/system/cmake_install.cmake:46 (include)
  /home/centos/smoketest/PR-14890/build/cmake_install.cmake:57 (include)
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:56 (file):
  file RPATH_CHANGE could not write new RPATH:
    /opt/HPCCSystems/lib
  to the file:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-sqs/libaws-cpp-sdk-sqs.so
  The current RPATH is:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-core:
  which does not contain:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-core:/home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/.deps/install/lib64:
  as was expected.
Call Stack (most recent call first):
  /home/centos/smoketest/PR-14890/build/system/cmake_install.cmake:46 (include)
  /home/centos/smoketest/PR-14890/build/cmake_install.cmake:57 (include)
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:63 (file):
  file RPATH_CHANGE could not write new RPATH:
    /opt/HPCCSystems/lib
  to the file:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-s3/libaws-cpp-sdk-s3.so
  The current RPATH is:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-core:
  which does not contain:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-core:/home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/.deps/install/lib64:
  as was expected.
Call Stack (most recent call first):
  /home/centos/smoketest/PR-14890/build/system/cmake_install.cmake:46 (include)
  /home/centos/smoketest/PR-14890/build/cmake_install.cmake:57 (include)
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:70 (file):
  file INSTALL cannot find
  /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/.deps/install/lib64/libaws-c-common.so.1.0.0.
Call Stack (most recent call first):
  /home/centos/smoketest/PR-14890/build/system/cmake_install.cmake:46 (include)
  /home/centos/smoketest/PR-14890/build/cmake_install.cmake:57 (include)
CPack Error: Error when generating package: hpccsystems-platform
make: *** [Makefile:130: package] Error 1
Package end (2021-04-23_14-18-19 5 sec
 ... ( comment is too long, 100 lines, 5663 bytes, truncated) ... 
[Build error:](https://gist.github.com/7744d1b709742ed921cfdf43d7804f17)
Build log [Build log](https://gist.github.com/HPCCSmoketest/7744d1b709742ed921cfdf43d7804f17#file-result-21-04-23-14-10-18-log)

```

jakesmith commented on esp/services/ws_dfu/ws_dfuService.cpp:
```diff
@@ -6030,6 +6030,10 @@ void CWsDfuEx::dFUFileAccessCommon(IEspContext &context, const CDfsLogicalFileNa
     getFileDafilesrvConfiguration(keyPairName, port, secure, fileName, groups);
+#ifdef _USE_OPENSSL
+    if (secure && keyPairName.isEmpty())
+        throw makeStringExceptionV(-1, "No keyPairName is found for %s in environment settings: /EnvSettings/Keys/ClusterGroup.", cluster.str());
```
@wangkx - here also

wangkx replied:
@jakesmith, now it is fixed.

HPCCSmoketest commented:
Process of PR-14890, label: h25498 starts now.
The reason of this test is: New commit.
Commit ID: d92e54bbe5a15022e8dbb5fd25e927b8280de3e8
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-91.ca-central-1.compute.internal

jakesmith reviewed (dismissed):
@wangkx - looks good

wangkx commented:
@richardkchapman this PR is ready to be merged.

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-91.ca-central-1.compute.internal
Sha: d92e54bbe5a15022e8dbb5fd25e927b8280de3e8
Build: success
Build: failed!
Number of Error(s): 12
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
warn(226) A package without a profile could throw errors or warnings. package: dojo-util
\terrors: 0
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:49 (file):
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:56 (file):
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:63 (file):
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:70 (file):
CPack Error: Error when generating package: hpccsystems-platform
make: *** [Makefile:130: package] Error 1
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:49 (file):
  file RPATH_CHANGE could not write new RPATH:
    /opt/HPCCSystems/lib
  to the file:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-core/libaws-cpp-sdk-core.so
  No valid ELF RPATH or RUNPATH entry exists in the file;
Call Stack (most recent call first):
  /home/centos/smoketest/PR-14890/build/system/cmake_install.cmake:46 (include)
  /home/centos/smoketest/PR-14890/build/cmake_install.cmake:57 (include)
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:56 (file):
  file RPATH_CHANGE could not write new RPATH:
    /opt/HPCCSystems/lib
  to the file:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-sqs/libaws-cpp-sdk-sqs.so
  The current RPATH is:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-core:
  which does not contain:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-core:/home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/.deps/install/lib64:
  as was expected.
Call Stack (most recent call first):
  /home/centos/smoketest/PR-14890/build/system/cmake_install.cmake:46 (include)
  /home/centos/smoketest/PR-14890/build/cmake_install.cmake:57 (include)
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:63 (file):
  file RPATH_CHANGE could not write new RPATH:
    /opt/HPCCSystems/lib
  to the file:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-s3/libaws-cpp-sdk-s3.so
  The current RPATH is:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-core:
  which does not contain:
    /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/aws-cpp-sdk-core:/home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/.deps/install/lib64:
  as was expected.
Call Stack (most recent call first):
  /home/centos/smoketest/PR-14890/build/system/cmake_install.cmake:46 (include)
  /home/centos/smoketest/PR-14890/build/cmake_install.cmake:57 (include)
CMake Error at /home/centos/smoketest/PR-14890/build/system/aws/cmake_install.cmake:70 (file):
  file INSTALL cannot find
  /home/centos/smoketest/PR-14890/build/system/aws/aws-sdk-cpp/.deps/install/lib64/libaws-c-common.so.1.0.0.
Call Stack (most recent call first):
  /home/centos/smoketest/PR-14890/build/system/cmake_install.cmake:46 (include)
  /home/centos/smoketest/PR-14890/build/cmake_install.cmake:57 (include)
CPack Error: Error when generating package: hpccsystems-platform
make: *** [Makefile:130: package] Error 1
Package end (2021-04-23_14-32-16 5 sec
 ... ( comment is too long, 100 lines, 5663 bytes, truncated) ... 
[Build error:](https://gist.github.com/9dbef39f5831236024a960a44a80eec9)
Build log [Build log](https://gist.github.com/HPCCSmoketest/9dbef39f5831236024a960a44a80eec9#file-result-21-04-23-14-24-19-log)

```

ghalliday commented:
@wangkx the PR now contains an invalid update to the aws submodule

wangkx commented:
@ghalliday not sure why I cannot see the invalid update in the PR. I will clean it.

jakesmith reviewed (commented):
@wangkx - will approve again when aws submodule issue resolved

jakesmith commented:
> @ghalliday not sure why I cannot see the invalid update in the PR. I will clean it.

![image](https://user-images.githubusercontent.com/902700/115889669-0bf4e280-a44c-11eb-9274-f17b631ab5da.png)

wangkx committed:
HPCC-25498 Check and send error messages in DFUFileAccessV2

1. Check the keyPairName setting in /EnvSettings/Keys/ClusterGroup.
2. Check the privateKey setting in /EnvSettings/Keys/KeyPair.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

wangkx commented:
@jakesmith the aws submodule issue is resolved

HPCCSmoketest commented:
Process of PR-14890, label: h25498 starts now.
The reason of this test is: Forced to re-test.
Commit ID: bd2881a4f6ed5bed7cdfb2aee52b24e1b1b408f8
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-205.ca-central-1.compute.internal

jakesmith reviewed (approved):
@wangkx - looks good

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-205.ca-central-1.compute.internal
Sha: bd2881a4f6ed5bed7cdfb2aee52b24e1b1b408f8
Build: success
Milestone:Install hpccsystems-platform-community_8.0.9-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |158 |158 |0 |0 |0 |57 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |24 sec  (00:00:24) |
|setup (thor) |9 |9 |0 |40 sec  (00:00:40) |
|setup (roxie) |19 |19 |0 |20 sec  (00:00:20) |
|test (hthor) |967 |967 |0 |900 sec  (00:15:00) |
|test (thor) |879 |879 |0 |1088 sec  (00:18:08) |
|test (roxie) |1038 |1038 |0 |908 sec  (00:15:08) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  599 sec (00:09:59) |  107 sec (00:01:47) |   21 sec (00:00:21) |   16 sec (00:00:16) | 1400 sec (00:23:20) |   19 sec (00:00:19) | 2171 sec (00:36:11) |

wangkx commented:
@richardkchapman this PR is ready to be merged.
